src/Common/CgroupsMemoryUsageObserver.cpp

```diff
@@ -0,0 +1,341 @@
+}
+
+void CgroupsMemoryUsageObserver::setLimits(uint64_t hard_limit_, uint64_t soft_limit_)
+{
+    if (hard_limit_ == hard_limit && soft_limit_ == soft_limit)
+        return;
+
+    stopThread();
+
+    hard_limit = hard_limit_;
+    soft_limit = soft_limit_;
+
+    on_hard_limit = [this, hard_limit_](bool up)
+    {
+        if (up)
+        {
+            LOG_WARNING(log, "Exceeded hard memory limit ({})", ReadableSize(hard_limit_));
+
+            /// Update current usage in memory tracker. Also reset free_memory_in_allocator_arenas to zero though we don't know if they are
+            /// really zero. Trying to avoid OOM ...
+            MemoryTracker::setRSS(hard_limit_, 0);
+        }
+        else
+        {
+            LOG_INFO(log, "Dropped below hard memory limit ({})", ReadableSize(hard_limit_));
+        }
+    };
+
+    on_soft_limit = [this, soft_limit_](bool up)
+    {
+        if (up)
+        {
+            LOG_WARNING(log, "Exceeded sort memory limit ({})", ReadableSize(soft_limit_));
+
+#if USE_JEMALLOC
+            LOG_INFO(log, "Purging jemalloc arenas");
+            mallctl("arena." STRINGIFY(MALLCTL_ARENAS_ALL) ".purge", nullptr, nullptr, nullptr, 0);
+#endif
+            /// Reset current usage in memory tracker. Expect zero for free_memory_in_allocator_arenas as we just purged them.
+            uint64_t current_usage = readMemoryUsage();
+            MemoryTracker::setRSS(current_usage, 0);
+
+            LOG_INFO(log, "Purged jemalloc arenas. Current memory usage is {}", ReadableSize(current_usage));
+        }
+        else
+        {
+            LOG_INFO(log, "Dropped below soft memory limit ({})", ReadableSize(soft_limit_));
+        }
+    };
+
+    startThread();
+
+    LOG_INFO(log, "Set new limits, soft limit: {}, hard limit: {}", ReadableSize(soft_limit_), ReadableSize(hard_limit_));
+}
+
+uint64_t CgroupsMemoryUsageObserver::readMemoryUsage() const
+{
+    return file.readMemoryUsage();
+}
+
+namespace
+{
+
+/// I think it is possible to mount the cgroups hierarchy somewhere else (e.g. when in containers).
+/// /sys/fs/cgroup was still symlinked to the actual mount in the cases that I have seen.
+const std::filesystem::path default_cgroups_mount = "/sys/fs/cgroup";
+
+/// Caveats:
+/// - All of the logic in this file assumes that the current process is the only process in the
+///   containing cgroup (or more precisely: the only process with significant memory consumption).
+///   If this is not the case, then other processe's memory consumption may affect the internal
+///   memory tracker ...
+/// - Cgroups v1 and v2 allow nested cgroup hierarchies. As v1 is deprecated for over half a
+///   decade and will go away at some point, hierarchical detection is only implemented for v2.
+/// - I did not test what happens if a host has v1 and v2 simultaneously enabled. I believe such
+///   systems existed only for a short transition period.
+
+std::optional<std::string> getCgroupsV2FileName()
```

Comment:
there are similar logic in `base/getMemoryAmount` to read V2File. The difference seems to be getting different sections of the file. Is it possible to reuse some code?

Reply:
Will do that, yes.